core/src/main/java/io/grpc/internal/BinaryLog.java

```diff
@@ -0,0 +1,208 @@
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Objects;
+import io.grpc.MethodDescriptor;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+
+/**
+ * A binary log class that is configured for a specific {@link MethodDescriptor}.
+ */
+public final class BinaryLog {
+  private static final Logger logger = Logger.getLogger(BinaryLog.class.getName());
+  private static final BinaryLog NOOP_LOG =
+      new BinaryLog(/*maxHeaderBytes=*/ 0, /*maxMessageBytes=*/ 0);
+  private final int maxHeaderBytes;
+  private final int maxMessageBytes;
+
+  @VisibleForTesting
+  BinaryLog(int maxHeaderBytes, int maxMessageBytes) {
+    this.maxHeaderBytes = maxHeaderBytes;
+    this.maxMessageBytes = maxMessageBytes;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (!(o instanceof BinaryLog)) {
+      return false;
+    }
+    BinaryLog that = (BinaryLog) o;
+    return this.maxHeaderBytes == that.maxHeaderBytes
+        && this.maxMessageBytes == that.maxMessageBytes;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(maxHeaderBytes, maxMessageBytes);
+  }
+
+  @Override
+  public String toString() {
+    return getClass().getSimpleName() + '['
+        + "maxHeaderBytes=" + maxHeaderBytes + ", "
+        + "maxMessageBytes=" + maxMessageBytes + "]";
+  }
+
+  private static final Factory DEFAULT_FACTORY =
+      new Factory(System.getenv("GRPC_BINARY_LOG_CONFIG"));
+
+  /**
+   * Accepts the fullMethodName and returns the binary log that should be used. The log may be
+   * a log that does nothing.
+   */
+  public static BinaryLog getLog(String fullMethodName) {
+    return DEFAULT_FACTORY.getLog(fullMethodName);
+  }
+
+  static final class Factory {
+    // '*' for global, 'service/*' for service glob, or 'service/method' for fully qualified.
+    private static final Pattern logPatternRe = Pattern.compile("[^{]+");
+    // A curly brace wrapped expression. Will be further matched with the more specified REs below.
+    private static final Pattern logOptionsRe = Pattern.compile("\\{[^}]+}");
+    private static final Pattern configRe = Pattern.compile(
+        String.format("^(%s)(%s)?$", logPatternRe.pattern(), logOptionsRe.pattern()));
+    // Regexes to extract per-binlog options
+    // The form: {m:256}
+    private static final Pattern msgRe = Pattern.compile("\\{m(?::(\\d+))?}");
+    // The form: {h:256}
+    private static final Pattern headerRe = Pattern.compile("\\{h(?::(\\d+))?}");
+    // The form: {h:256,m:256}
+    private static final Pattern bothRe = Pattern.compile("\\{h(?::(\\d+))?;m(?::(\\d+))?}");
+
+    private final boolean enabled;
+    private final Map<String, BinaryLog> perMethodLogs = new HashMap<String, BinaryLog>();
```

Comment:
Make these unmodifiable? At the very least it'd be nice to assign them at the end of the constructor, so it's more obvious no code is modifying them after they're set.

Reply:
Switched to an unmodifiable map